lib/index.d.ts

```diff
@@ -0,0 +1,271 @@
+ * urllib.request('http://httptest.cnodejs.net/test/post', args, function(err, data, res) {});
+ *
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function request<T = any>(url: string | url.URL, options?: RequestOptions): Promise<HttpClientResponse<T>>;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function request<T = any>(url: string | url.URL, callback: Callback<T>): void;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function request<T = any>(url: string | url.URL, options: RequestOptions, callback: Callback<T>): void;
+
+/**
+ * Handle request with a callback.
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function requestWithCallback<T = any>(url: string | url.URL, callback: Callback<T>): void;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function requestWithCallback<T = any>(url: string | url.URL, options: RequestOptions, callback: Callback<T>): void;
+
+/**
+ * yield urllib.requestThunk(url, args)
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function requestThunk<T = any>(url: string | url.URL, options?: RequestOptions): (callback: Callback<T>) => void;
+
+/**
+ * alias to request.
+ * Handle all http request, both http and https support well.
+ *
+ * @example
+ * // GET http://httptest.cnodejs.net
+ * urllib.request('http://httptest.cnodejs.net/test/get', function(err, data, res) {});
+ * // POST http://httptest.cnodejs.net
+ * var args = { type: 'post', data: { foo: 'bar' } };
+ * urllib.request('http://httptest.cnodejs.net/test/post', args, function(err, data, res) {});
+ *
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ * @param options Optional, @see RequestOptions.
+ */
+export function curl<T = any>(url: string | url.URL, options?: RequestOptions): Promise<HttpClientResponse<T>>;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function curl<T = any>(url: string | url.URL, callback: Callback<T>): void;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ */
+export function curl<T = any>(url: string | url.URL, options: RequestOptions, callback: Callback<T>): void;
+/**
+ * The default request timeout(in milliseconds).
+ */
+export const TIMEOUT: number;
+/**
+ * The default request & response timeout(in milliseconds).
+ */
+export const TIMEOUTS: [number, number];
+
+/**
+ * Request user agent.
+ */
+export const USER_AGENT: string;
+
+/**
+ * Request http agent.
+ */
+export const agent: http.Agent;
+
+/**
+ * Request https agent.
+ */
+export const httpsAgent: https.Agent;
+
+export class HttpClient<O = RequestOptions> extends EventEmitter {
```

Comment:
```suggestion
export class HttpClient<O extends RequestOptions = RequestOptions> extends EventEmitter {
```

限制一下 O 必须得 extends RequestOptions 好一些吧